File: app/src/main/java/com/riprog/launcher/ThemeMechanism.java
package com.riprog.launcher;

import android.app.UiModeManager;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.RippleDrawable;
import android.os.Build;
import android.util.TypedValue;
import android.widget.LinearLayout;

public class ThemeMechanism {

    public static void applyThemeMode(Context context, String mode) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            UiModeManager uiModeManager = (UiModeManager) context.getSystemService(Context.UI_MODE_SERVICE);
            int nightMode = UiModeManager.MODE_NIGHT_AUTO;
            if ("light".equals(mode)) nightMode = UiModeManager.MODE_NIGHT_NO;
            else if ("dark".equals(mode)) nightMode = UiModeManager.MODE_NIGHT_YES;

            if (uiModeManager.getNightMode() != nightMode) {
                uiModeManager.setApplicationNightMode(nightMode);
            }
        }
    }

    public static Integer getSystemAccentColor(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            try {
                return context.getResources().getColor(android.R.color.system_accent1_400, context.getTheme());
            } catch (Exception ignored) {
                return null;
            }
        }
        return null;
    }

    public static void applySettingItemStyle(Context context, LinearLayout item, SettingsManager settingsManager) {
        item.setClickable(true);
        item.setFocusable(true);

        float radius = dpToPx(context, 12);
        boolean isNight = (context.getResources().getConfiguration().uiMode & android.content.res.Configuration.UI_MODE_NIGHT_MASK) ==
                android.content.res.Configuration.UI_MODE_NIGHT_YES;

        GradientDrawable shape = new GradientDrawable();
        int baseColor;
        if (settingsManager.isLiquidGlass()) {
            baseColor = isNight ? 0x26FFFFFF : 0x1A000000;
        } else {
            baseColor = isNight ? 0x1AFFFFFF : 0x0D000000;
        }
        shape.setColor(baseColor);
        shape.setCornerRadius(radius);
        if (settingsManager.isLiquidGlass()) {
            shape.setStroke(dpToPx(context, 1), 0x20FFFFFF);
        }

        GradientDrawable mask = new GradientDrawable();
        mask.setColor(Color.BLACK);
        mask.setCornerRadius(radius);

        int rippleColor;
        try {
            rippleColor = context.getColor(R.color.search_background);
        } catch (Exception e) {
            rippleColor = 0x40FFFFFF;
        }

        item.setBackground(new RippleDrawable(
            ColorStateList.valueOf(rippleColor),
            shape,
            mask
        ));
    }

    private static int dpToPx(Context context, int dp) {
        return (int) TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP, (float) dp, context.getResources().getDisplayMetrics()
        );
    }
}

File: app/src/main/java/com/riprog/launcher/ui/common/ThemeMechanism.kt
package com.riprog.launcher.ui.common

import android.app.UiModeManager
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.RippleDrawable
import android.os.Build
import android.util.TypedValue
import android.widget.LinearLayout
import com.riprog.launcher.data.local.prefs.LauncherPreferences

/**
 * Theme switching and dynamic color mechanisms.
 */
object ThemeMechanism {

    /**
     * Applies the selected theme mode (system, light, dark).
     */
    fun applyThemeMode(context: Context, mode: String?) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val uiModeManager = context.getSystemService(Context.UI_MODE_SERVICE) as UiModeManager
            val nightMode = when (mode) {
                "light" -> UiModeManager.MODE_NIGHT_NO
                "dark" -> UiModeManager.MODE_NIGHT_YES
                else -> UiModeManager.MODE_NIGHT_AUTO
            }
            if (uiModeManager.nightMode != nightMode) {
                uiModeManager.setApplicationNightMode(nightMode)
            }
        }
    }

    /**
     * Extracts dynamic accent color from Android 12+ system colors.
     */
    fun getSystemAccentColor(context: Context): Int? {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            return try {
                context.resources.getColor(android.R.color.system_accent1_400, context.theme)
            } catch (ignored: Exception) {
                null
            }
        }
        return null
    }

    /**
     * Applies the specialized "Liquid Glass" item style to a container.
     */
    fun applySettingItemStyle(context: Context, item: LinearLayout, settingsManager: LauncherPreferences) {
        item.isClickable = true
        item.isFocusable = true

        val radius = dpToPx(context, 12).toFloat()
        val isNight = (context.resources.configuration.uiMode and android.content.res.Configuration.UI_MODE_NIGHT_MASK) ==
                android.content.res.Configuration.UI_MODE_NIGHT_YES

        val shape = GradientDrawable()
        val baseColor = if (settingsManager.isLiquidGlass) {
            if (isNight) 0x26FFFFFF else 0x1A000000
        } else {
            if (isNight) 0x1AFFFFFF else 0x0D000000
        }
        shape.setColor(baseColor)
        shape.cornerRadius = radius
        if (settingsManager.isLiquidGlass) {
            shape.setStroke(dpToPx(context, 1), 0x20FFFFFF)
        }

        val mask = GradientDrawable()
        mask.setColor(Color.BLACK)
        mask.cornerRadius = radius

        // Note: R.color.search_background should be defined in colors.xml
        val rippleColor = try {
            context.getColor(context.resources.getIdentifier("search_background", "color", context.packageName))
        } catch (e: Exception) {
            0x40FFFFFF
        }

        item.background = RippleDrawable(
            ColorStateList.valueOf(rippleColor),
            shape,
            mask
        )
    }

    private fun dpToPx(context: Context, dp: Int): Int {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP, dp.toFloat(), context.resources.displayMetrics
        ).toInt()
    }
}
